### PR TITLE
Adds Odin read-only query tools

### PR DESCRIPTION
**Pull request by akshaydeo**

### CODE REVIEW

**Comment by akshaydeo** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6214?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#7382** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7382?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7316** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7316?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7418** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7418?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7186** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7186?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7183** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7183?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7151** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7151?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7086** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7086?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7070** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7070?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7068** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7068?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6882** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6882?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6852** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6852?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6851** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6851?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6850** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6850?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6849** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6849?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6848** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6848?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6847** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6847?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6221** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6221?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6220** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6220?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6219** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6219?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6218** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6218?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6217** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6217?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6216** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6216?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6215** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6215?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6214** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6214?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6214?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#6213** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6213?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6212** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6212?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6207** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6207?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#6227](https://github.com/maximhq/bifrost/pull/6227) <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6227?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#6196** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6196?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6195** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/6195?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `dev`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: transports/bifrost-http/handlers/odintools.go [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+		return fmt.Sprintf(`{"error":"could not serialize result: %s"}`, err.Error())
+	}
+	if len(encoded) <= odinMaxToolResultBytes {
+		return encoded
+	}
+	return fmt.Sprintf(
+		`{"error":"result too large (%d bytes, limit %d). Narrow the time range, add filters, or lower the limit, then try again.","truncated":true}`,
+		len(encoded), odinMaxToolResultBytes,
+	)
+}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/odintools.go]: This is a stack-slicing artifact rather than a gap: #6214 adds the tool definitions and #6215 adds the agent loop that runs them. On #6215 the production path wires all three - `framework/warp/agent.go:92` builds the tool set with `buildTools()`, line 129 declares them to the provider via `chatTools()`, and line 269 passes every result through `boundToolResult()`, so the 16 KiB ceiling is enforced on the production path. Splitting them the other way round would have meant landing an executor with no tools to execute.

File: transports/bifrost-http/handlers/odintools_test.go [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+func TestOdinMetricsRejectsTooManyBuckets(t *testing.T) {
+	_, err := runOdinTool(t, "query_metrics", &odinToolDeps{logManager: &fakeOdinLogManager{}}, map[string]any{
+		// A minute-scale bucket over a long window is what blows the count up.
+		"filters": map[string]any{"start_time": "-47h", "end_time": "2026-08-17T00:00:00Z"},
+		"metrics": []any{"cost"},
+	})
+	if err != nil {
+		require.ErrorContains(t, err, "buckets")
+	}
+}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/odintools_test.go]: Fixed in this branch. `TestWarpMetricsRejectsTooManyBuckets` now pins `Now` and uses two absolute timestamps rather than a relative window, and the assertion is unconditional. Worth noting why the original was vacuous: `DefaultBucketSize` is adaptive and self-caps every band, so `MaxHistogramBuckets` is only reachable across a span of roughly 16 years - the test now uses about 26.

File: transports/bifrost-http/handlers/odintools.go [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+
+func truncateOdinText(text string, limit int) string {
+	if len(text) <= limit {
+		return text
+	}
+	return text[:limit] + "... [truncated]"
+}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/odintools.go]: Fixed in this branch. `truncateText` in `framework/warp/tools.go` counts with `utf8.RuneCountInString` and slices only at rune-boundary indices from `for i := range text`, so a multi-byte character can no longer be cut in half. Pinned by `TestWarpTruncateTextCutsOnRuneBoundaries`.

File: framework/warp/flows.go [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+			stopReasons, err := deps.logManager.GetAvailableStopReasons(ctx, limit, query)
+			if err != nil {
+				return nil, fmt.Errorf("could not list stop reasons: %w", err)
+			}
+			return map[string]any{
+				"models":       models,
+				"virtual_keys": virtualKeys,
+				"apps":         apps,
+				"stop_reasons": stopReasons,
+			}, nil
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/flows.go:436]: Fixed in this branch. The `describe_filter_space` description in `framework/warp/flows.go` now lists only what the executor actually returns - models, virtual keys, apps and stop reasons - with a comment recording why the provider claim was removed. Pinned by `TestWarpDescribeFilterSpaceDescriptionMatchesResult`.

File: framework/warp/tools_test.go [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+	Now = func() time.Time { return time.Date(2026, 8, 17, 0, 0, 0, 0, time.UTC) }
+	defer func() { Now = previous }()
+	_, err := runTool(t, "query_metrics", &ToolDeps{logManager: &fakeLogReader{}}, map[string]any{
+		// A minute-scale bucket over a long window is what blows the count up.
+		"filters": map[string]any{"start_time": "-47h", "end_time": "2026-08-17T00:00:00Z"},
+		"metrics": []any{"cost"},
+	})
+	if err != nil {
+		require.ErrorContains(t, err, "buckets")
+	}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools_test.go]: Fixed in this branch. `TestWarpMetricsRejectsTooManyBuckets` calls `require.Error` and then `require.ErrorContains(t, err, "buckets")`, so it fails rather than passes when no error is returned.

File: framework/warp/flows.go [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+			for _, metric := range metrics {
+				switch metric {
+				case "summary":
+					stats, err := deps.logManager.GetStats(ctx, filters)
+					if err != nil {
+						return nil, fmt.Errorf("stats query failed: %w", err)
+					}
+					out["summary"] = stats
+				case "requests":
+					result, err := deps.logManager.GetHistogram(ctx, filters, bucket)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/flows.go:186]: Fixed in this branch. `framework/warp/flows.go` returns an explicit error for `group_by: "provider"` on the `requests` metric, naming the metrics that do support it. Pinned by `TestWarpMetricsRejectsUnsupportedProviderGrouping`.

File: framework/warp/tools.go [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+    "team_ids": {"type": "array", "items": {"type": "string"}},
+    "customer_ids": {"type": "array", "items": {"type": "string"}},
+    "user_ids": {"type": "array", "items": {"type": "string"}},
+    "business_unit_ids": {"type": "array", "items": {"type": "string"}},
+    "apps": {"type": "array", "items": {"type": "string"}},
+    "min_latency": {"type": "number", "description": "Milliseconds."},
+    "max_latency": {"type": "number", "description": "Milliseconds."},
+    "min_cost": {"type": "number"},
+    "max_cost": {"type": "number"},
+    "content_search": {"type": "string", "description": "Substring match against request and response content."}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools.go]: Fixed in this branch. `FilterSchema` sets `maxItems: 50` and `maxLength: 500`, and the bounds are enforced in code rather than only advertised: `MaxFilterValues` in `stringSliceField` and `MaxContentSearchChars` on `content_search`. Pinned by `TestWarpFilterBoundsInputSize`.

File: framework/warp/tools.go [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+	filters.VirtualKeyIDs = stringSlice(raw["virtual_key_ids"])
+	filters.TeamIDs = stringSlice(raw["team_ids"])
+	filters.CustomerIDs = stringSlice(raw["customer_ids"])
+	filters.UserIDs = stringSlice(raw["user_ids"])
+	filters.BusinessUnitIDs = stringSlice(raw["business_unit_ids"])
+	filters.Apps = stringSlice(raw["apps"])
+	filters.MinLatency = floatPtr(raw["min_latency"])
+	filters.MaxLatency = floatPtr(raw["max_latency"])
+	filters.MinCost = floatPtr(raw["min_cost"])
+	filters.MaxCost = floatPtr(raw["max_cost"])
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools.go]: Fixed in this branch. `parseFilters` routes every value through `stringSliceField`/`floatField`, which return an error on a non-array, a non-string element or a non-number instead of coercing. Silently dropping a malformed value was the dangerous half: it broadened the query rather than failing it. Pinned by `TestWarpFilterRejectsWrongShapedValues`.

File: framework/warp/flows.go [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+			filters, err := filterArg(args, now)
+			if err != nil {
+				return nil, err
+			}
+			metrics := stringSlice(args["metrics"])
+			if len(metrics) == 0 {
+				return nil, fmt.Errorf("metrics must list at least one of: summary, requests, tokens, cost, latency, throughput")
+			}
+			groupBy, _ := args["group_by"].(string)
+			byProvider := groupBy == "provider"
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/flows.go]: Fixed. `groupByProvider` in `framework/warp/tools.go` now accepts only an omitted value, `""`, `"none"` or `"provider"`, and errors on anything else (including a non-string). You identified the real shape of this: the schema is advertised to the model and never enforced on the arguments that come back, so comparing against `"provider"` alone let every other value fall through to the ungrouped branch - the model asked for one breakdown and got aggregates for another, with nothing marking it as a different answer.

Pinned by `TestWarpMetricsRejectsUnknownGrouping`, which rejects `model`, `virtual_key`, `PROVIDER` and `" provider"` and asserts all four accepted forms still work. Making the accepted `provider` case actually run also exposed a gap in the test fake: `LogReaderStub` embeds a nil `LogReader`, so every unimplemented method panics, which is why the grouped branch had only ever been asserted on its rejection. `fakeLogReader` now serves `GetProviderCostHistogram`.

File: framework/warp/tools.go [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+
+func boolArg(args map[string]any, key string) bool {
+	value, _ := args[key].(bool)
+	return value
+}
+
+func filterArg(args map[string]any, now time.Time) (*logstore.SearchFilters, error) {
+	raw, _ := args["filters"].(map[string]any)
+	return parseFilters(raw, now)
+}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools.go:434]: Fixed. `filtersObject` in `framework/warp/tools.go` reads the top-level argument and errors when it is present but not an object, naming the type it got. Absent and nil still legitimately mean no filters. As you noted this is only the top-level shape - the fields inside a valid object stay `parseFilters`' job.

The direction of the failure is what made it worth fixing: a discarded assertion turned a malformed filter into the default unfiltered 24-hour query, so a bad argument widened the result instead of failing it and nothing told the model its filter had been dropped. Pinned by `TestWarpFilterRejectsNonObjectFilters` across a string, an array, a number and a bool, with absent and `{}` still passing.

File: transports/bifrost-http/handlers/logging.go [hunk 11]
```diff
@@ -1106,30 +1106,14 @@ func (h *LoggingHandler) getLogsHistogram(ctx *fasthttp.RequestCtx) {
...
-	case duration >= 3*24*time.Hour: // >= 3 days
-		return 8 * 3600 // 8 hours
-	case duration >= 24*time.Hour: // >= 24 hours
-		return 3600 // Hourly
-	case duration >= 2*time.Hour: // >= 2 hours
-		return 600 // 10 minutes
-	default:
-		return 60 // 1 minute buckets for < 2 hours
-	}
+	return logstore.DefaultBucketSize(start, end)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/logging.go:1116]: Declining - this is the fourth report of the same `go.mod` reading, so let me give the evidence rather than just the conclusion.

`DefaultBucketSize` is defined in this repo at `framework/logstore/bucketsize.go:14`, introduced by this very PR. The pin in `transports/go.mod` is not what the build resolves: `go.work` is gitignored (`.gitignore:39`) and regenerated by `.github/workflows/scripts/setup-go-workspace.sh` before any build, so `transports` compiles against the in-repo `framework`, not a published release.

You can see the same thing from the other direction: if the pinned version governed, every cross-module PR in this repo would fail the same way, including the ones already merged. Reading `go.mod` alone will keep producing this finding on any PR that adds a framework symbol and uses it from transports.

File: transports/bifrost-http/handlers/odinflows.go [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+			stopReasons, err := deps.logManager.GetAvailableStopReasons(ctx, limit, query)
+			if err != nil {
+				return nil, fmt.Errorf("could not list stop reasons: %w", err)
+			}
+			return map[string]any{
+				"models":       models,
+				"virtual_keys": virtualKeys,
+				"apps":         apps,
+				"stop_reasons": stopReasons,
+			}, nil
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/odinflows.go]: Already fixed. The file is now `framework/warp/flows.go` (the handler split out during the Odin-to-Warp rename), and `describe_filter_space` names exactly what it returns: "models, virtual keys, apps and stop reasons". Providers were removed from the description rather than added to the result, with the reason recorded above it - `LogReader` has no listing method for providers, so advertising the dimension would cause the very failure the tool exists to prevent. `stop_reasons` is now documented.

File: framework/warp/flows.go [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+			}
+			limit := intArg(args, "limit", 10, MaxLogRows)
+			sortBy, _ := args["sort_by"].(string)
+			if sortBy == "" {
+				sortBy = "timestamp"
+			}
+			order, _ := args["order"].(string)
+			if order == "" {
+				order = "desc"
+			}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/flows.go]: Already fixed. Both values go through `enumArg`: `enumArg(args, "sort_by", "timestamp", []string{"timestamp", "latency", "tokens", "cost"})` and `enumArg(args, "order", "desc", []string{"asc", "desc"})`, each returning an error for a value outside the set rather than letting `searchLogs` coerce it. An unsupported `sort_by` can no longer run a timestamp-ordered query in silence.

File: framework/warp/tools.go [hunk 14]
```diff
@@ -0,0 +1,555 @@
...
+
+// parseTime accepts RFC3339 or a relative offset like "-7d". Models reach
+// for relative offsets constantly ("last week"), and making them compute an
+// absolute timestamp from a date they only half-know is a reliable source of
+// wrong answers.
+func parseTime(value any, now time.Time) (*time.Time, error) {
+	text, ok := value.(string)
+	if !ok || strings.TrimSpace(text) == "" {
+		return nil, nil
+	}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools.go]: Already fixed. `parseTime` returns `nil, nil` only for a `nil` value; a present non-string returns "must be an RFC3339 timestamp or a relative offset like -7d, got %T" and a blank string returns its own error. The comment above it records exactly the failure you describe: a malformed argument became a valid query over different dates and the answer came back looking like the one that was asked for.

File: framework/warp/tools.go [hunk 15]
```diff
@@ -0,0 +1,591 @@
...
+	text = strings.TrimSpace(text)
+	if strings.HasPrefix(text, "-") {
+		// time.ParseDuration has no day unit, which is the one people actually use.
+		if strings.HasSuffix(text, "d") {
+			var days float64
+			if _, err := fmt.Sscanf(text, "-%fd", &days); err == nil && days > 0 {
+				result := now.Add(-time.Duration(days * float64(24*time.Hour)))
+				return &result, nil
+			}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools.go:253]: Fixed as prescribed. `fmt.Sscanf(text, "-%fd", ...)` is gone - it matched the `-7d` prefix of `-7daysd` and ignored the rest. The value is now the text between the leading `-` and the final `d`, parsed with `strconv.ParseFloat`, and rejected unless it is finite and positive: `strconv.ParseFloat(strings.TrimSuffix(text[1:], "d"), 64)` with explicit `math.IsNaN` / `math.IsInf` / `<= 0` guards. Table-driven test covers `-7daysd`, `-7d-extrad`, `-d`, `-0d`, `--7d`, `-NaNd` and `-Infd` as rejected, and `-7d`, `-1d`, `-0.5d`, `-30d` as correct durations.

File: framework/warp/tools.go [hunk 16]
```diff
@@ -0,0 +1,591 @@
...
+func stringSlice(value any) []string {
+	items, ok := value.([]any)
+	if !ok {
+		return nil
+	}
+	result := make([]string, 0, len(items))
+	for _, item := range items {
+		if text, ok := item.(string); ok && text != "" {
+			result = append(result, text)
+		}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools.go]: Fixed as suggested, before the metric loop. `stringSlice` is deleted - `query_metrics` was its only caller - and replaced by `enumSliceArg(args, "metrics", "metric", ...)`, which rejects a non-string element by index ("metrics[1] must be a string, got int") and an unknown value by index ("unknown metric at metrics[0]"), instead of dropping either. `["cost", 42]` is now an error rather than a successful answer about cost alone.

The `default:` case in the metric switch is kept deliberately, now unreachable through this path, as a drift guard: a metric added to the allowed list but not to the switch would otherwise be accepted and silently produce nothing. That is noted in a comment so it does not read as dead code.

File: framework/warp/tools.go [hunk 17]
```diff
@@ -0,0 +1,591 @@
...
+	if !present || value == nil {
+		return nil, nil
+	}
+	items, ok := value.([]any)
+	if !ok {
+		return nil, fmt.Errorf("%s must be an array of strings", key)
+	}
+	if len(items) > MaxFilterValues {
+		return nil, fmt.Errorf("%s lists %d values; at most %d are accepted. Narrow by dimension instead", key, len(items), MaxFilterValues)
+	}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools.go:332]: Fixed as suggested. `stringSliceField` now returns `"%s[%d] must not be empty"` instead of skipping the element, so `models: [""]` is a validation error naming the field and the index rather than an unfiltered query. Trailing-whitespace-only elements are caught too. The all-empty case previously collapsed to `nil`, which is what made the filter disappear entirely.

File: framework/warp/tools.go [hunk 18]
```diff
@@ -0,0 +1,591 @@
...
+	// Clamp rather than reject. The cap exists to protect the context window,
+	// not to police the model, and failing the call would just cost another
+	// round trip to arrive at the number we would have used anyway.
+	return min(result, max)
+}
+
+func boolArg(args map[string]any, key string) bool {
+	value, _ := args[key].(bool)
+	return value
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools.go]: Fixed as suggested. `boolArg` returns `(bool, error)`: absent or nil is `false, nil`, and a present non-boolean is `"%s must be a boolean, got %T"`. Both callers in `flows.go` propagate it - `include_content` in `query_logs` and `include_performance` in the model rankings flow. A malformed flag can no longer produce a successful answer with the requested content quietly missing, which is indistinguishable from a deployment that stores none.

File: framework/warp/flows.go [hunk 19]
```diff
@@ -0,0 +1,402 @@
...
+  },
+  "required": ["filters"]
+}`,
+		execute: func(ctx context.Context, deps *ToolDeps, args map[string]any) (any, error) {
+			now := Now()
+			filters, err := filterArg(args, now)
+			if err != nil {
+				return nil, err
+			}
+			limit := intArg(args, "limit", 10, MaxLogRows)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/flows.go]: Fixed as prescribed. `intArg` now returns `(int, error)`: absent or nil is still the default, a non-number is rejected by type, a fractional value is rejected rather than truncated, and anything below 1 is an error instead of a silent fallback. The upper-bound clamp is unchanged and still clamps rather than failing - the cap protects the context window, and failing there would only cost a round trip to reach the number we would have used anyway.

All four call sites propagate the error. The fourth is worth noting: `semantic_search_logs` on a later branch in this stack calls `intArg(args, "limit", 0, MaxLogRows)` where fallback 0 means "use the configured semantic limit", so absent still resolves to 0 there rather than to a row count. The restack surfaced that one - a per-branch vet sweep caught it as `multiple-value intArg ... in single-value context` before it could reach anybody.

File: framework/warp/flows.go [hunk 20]
```diff
@@ -0,0 +1,402 @@
...
+	}
+}
+
+// ------------------------------------------------------------- flow 2: metrics
+
+func queryMetricsTool() Tool {
+	return Tool{
+		name: "query_metrics",
+		description: "Aggregate statistics and time series over requests: totals, cost, tokens, latency percentiles, throughput. " +
+			"This is the cheapest way to answer 'how much', 'how many' and 'is it getting worse'. Series are returned summarized (total, mean, min, max, first, last) rather than bucket by bucket. " +
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/flows.go]: Fixed by correcting the description rather than by projecting the payload. There is no summarizer anywhere in the package - the histogram results are assigned straight to `out` and serialized with their `Buckets` arrays, so "returned summarized (total, mean, min, max, first, last)" was simply describing something the code never did.

The description now says the series come back bucket by bucket, with the bucket width chosen from the window so the count stays small - bounded rather than summarized - and tells the model to read the buckets instead of expecting a precomputed total. I chose this over projecting to summary fields because that would discard the shape of the series, which is exactly what "is it getting worse" questions need.

File: framework/warp/flows.go [hunk 21]
```diff
@@ -0,0 +1,402 @@
...
+			}
+			byProvider, err := groupByProvider(args)
+			if err != nil {
+				return nil, err
+			}
+
+			bucket, err := bucketSize(filters)
+			if err != nil {
+				return nil, err
+			}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/flows.go]: Fixed as suggested. The bucket is computed only when a histogram metric was actually requested: `if slices.ContainsFunc(metrics, isHistogramMetric)`. A new `isHistogramMetric` helper records the rule - only `summary` returns no buckets; every other metric calls one of the histogram readers.

Running `bucketSize` unconditionally meant a valid long-range `metrics: ["summary"]` failed on a bucket-count limit before `GetStats` was ever called - a limit on a series the answer does not contain.

File: framework/warp/flows.go [hunk 22]
```diff
@@ -0,0 +1,402 @@
...
+		description: "List the values that actually appear in this deployment's logs - models, virtual keys, apps and stop reasons. " +
+			"Call this before filtering by a name you are not certain about. Guessing a model or key name returns an empty result that looks like a real answer.",
+		schemaJSON: `{
+  "type": "object",
+  "properties": {
+    "search": {"type": "string", "description": "Optional substring to narrow the returned values."}
+  }
+}`,
+		execute: func(ctx context.Context, deps *ToolDeps, args map[string]any) (any, error) {
+			query, _ := args["search"].(string)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/flows.go]: Fixed as suggested. `describe_filter_space` now rejects a present non-string `search` with `"search must be a string, got %T"` instead of discarding the type assertion. The discarded assertion turned a malformed value into `""`, and the tool then ran four unfiltered discovery queries and returned every value in the deployment - a broader answer than the one asked for, with nothing in the payload to say it had been widened. An absent or null `search` still means "no filter", which is the documented behaviour.

File: framework/warp/tools.go [hunk 23]
```diff
@@ -0,0 +1,627 @@
...
+	}
+	if end == nil {
+		end = &now
+	}
+	if start == nil {
+		defaulted := end.Add(-DefaultLookback)
+		start = &defaulted
+	}
+	if start.After(*end) {
+		return nil, fmt.Errorf("start_time must be before end_time")
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools.go:172]: Fixed as prescribed, at the parser boundary. `parseFilters` checks presence for `start_time` and `end_time` before calling `parseTime`, and rejects a present `nil` with `"must be a string, got null; omit the field to use the default window"`. Indexing the map gave `nil` for an absent key and an explicit JSON `null` alike, so `"start_time": null` quietly took the default window - and `FilterSchema` declares both as strings, so `null` was never in the contract. Omitting the field is still how the default is requested.

This is the one of the six I could show a genuine red for: removing just this guard made the new test fail with "An error is expected but got nil". The others change function signatures, so reverting them breaks compilation rather than producing a real failure - see the note on the `intArg` thread.

File: framework/warp/tools.go [hunk 24]
```diff
@@ -0,0 +1,627 @@
...
+func enumSliceArg(args map[string]any, key, noun string, allowed []string) ([]string, error) {
+	value, present := args[key]
+	if !present || value == nil {
+		return nil, fmt.Errorf("%s must list at least one of: %s", key, strings.Join(allowed, ", "))
+	}
+	items, ok := value.([]any)
+	if !ok {
+		return nil, fmt.Errorf("%s must be an array of strings", key)
+	}
+	result := make([]string, 0, len(items))
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools.go:293]: Fixed as suggested, plus the case it opens up. `enumSliceArg` now rejects a list longer than the allowed vocabulary, since each metric is a separate database query and a list bounded only by "every element is valid" is not bounded at all - 200 copies of `"cost"` was 200 sequential queries, every one of them individually legitimate.

It also rejects a repeated value now: `"metrics[1] repeats \"cost\"; each value runs its own query, so asking twice only costs twice"`. A length cap alone would still have allowed six duplicates, so the dedup is what actually makes the work proportional to the question. Both are covered by a table-driven test alongside the type and unknown-value cases.

File: framework/warp/tools.go [hunk 25]
```diff
@@ -0,0 +1,665 @@
...
+	}
+	if value, present := raw["content_search"]; present && value != nil {
+		search, ok := value.(string)
+		if !ok {
+			return nil, fmt.Errorf("content_search must be a string")
+		}
+		if len(search) > MaxContentSearchChars {
+			return nil, fmt.Errorf("content_search is %d characters; at most %d are accepted", len(search), MaxContentSearchChars)
+		}
+		filters.ContentSearch = search
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools.go]: Fixed as suggested, both halves. `FilterSchema` gives `content_search` `minLength: 1` and tells the model to omit the field rather than send an empty string, and `parseFilters` now rejects a present `null`, an empty string, and a whitespace-only value with messages that say to omit the field instead.

The silent-widening you describe is the reason: the logstore applies the content predicate only when `ContentSearch` is non-empty, so an empty value reached the query as if the field had been omitted and a filtered question came back with unfiltered rows, with nothing to say the filter had been dropped. Covered by a test over `""`, `"   "` and `nil`, plus the omitted and valid cases.

File: framework/warp/flows.go [hunk 26]
```diff
@@ -0,0 +1,439 @@
...
+  "properties": {
+    "log_id": {"type": "string"}
+  },
+  "required": ["log_id"]
+}`,
+		execute: func(ctx context.Context, deps *ToolDeps, args map[string]any) (any, error) {
+			id, _ := args["log_id"].(string)
+			if id == "" {
+				return nil, fmt.Errorf("log_id is required")
+			}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/flows.go:93]: Fixed as suggested. `log_id` now goes through a new `stringArg` helper, matching how every other argument in the file is read. A present non-string became `""` and the tool answered "log_id is required", so the model was told it had omitted a field it actually sent - and retried with the same wrong shape. `stringArg` rejects a non-string by type, a null, and an empty or whitespace-only value, each with a message naming what was wrong.

File: framework/warp/tools.go [hunk 27]
```diff
@@ -0,0 +1,677 @@
...
+		// was empty, so models: [""] ran an unfiltered query and returned a
+		// broader result that reads as an answer to the narrow question.
+		if strings.TrimSpace(text) == "" {
+			return nil, fmt.Errorf("%s[%d] must not be empty", key, index)
+		}
+		result = append(result, text)
+	}
+	if len(result) == 0 {
+		return nil, nil
+	}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/warp/tools.go:349]: Fixed as prescribed, both halves. The nine filter-array schemas now carry `minItems: 1` and `"items": {"type": "string", "minLength": 1}`, and `stringSliceField` rejects a present `null` and an explicitly empty array at runtime with messages pointing at omitting the field instead.

Your reasoning is exactly the failure: `applyFilters` adds a predicate only for a non-empty slice, so both shapes reached the logstore as if the field had been omitted and the query ran unfiltered - a broader answer than the one asked for, with nothing in the payload to say the filter had been dropped.

Confirmed red first by removing just those two guards: "An error is expected but got nil" on the empty-array case.

**Comment by Pratham-Mishra04** [comment]
### Merge activity

* **Sep 23, 12:23 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/6214).
